### PR TITLE
Optimize committee helpers

### PR DESCRIPTION
**Pull request by terencechain**
A lot of places in our code `chainService.HeadState` was taken for granted. The call gets more expensive as state size grows and more validators entered testnet. This deficiency was optimized by providing `chainServce.HeadValidator` and `chainServiceSeed`. Then a few refactor of current committee helpers API to allow indices and seed instead of state object. 

Change list:
* Update `BeaconCommittee` API 
from `(beaconState, att1.Data.Slot, att1.Data.CommitteeIndex)`
to `(activeValidatorIndices, seed, att1.Data.Slot, att1.Data.CommitteeIndex)`
* Update `CommitteeCountAtSlot` API 
from `(state *pb.BeaconState, slot uint64)` 
to `(validatorCount uint64)`
* Update `IsAggregator` API 
from `(state *pb.BeaconState, slot uint64, index uint64, slotSig []byte)` 
to `(committeeCount uint64, slot uint64, index uint64, slotSig []byte)`
* Add `HeadValidator` to chains service info interface
* Add `HeadSeed` to chain service info interface

400x improvements running a beacon node with mainnet config 

Screen Shot 2019-12-18 at 5 50 54 PM

Screen Shot 2019-12-18 at 6 12 40 PM

Will open subsequent PRs to clean up rest of code with this change

### CODE REVIEW

File: beacon-chain/core/helpers/committee.go [hunk 1]
```diff
@@ -29,23 +34,22 @@ var committeeCache = cache.NewCommitteesCache()
...
 //        len(get_active_validator_indices(state, epoch)) // SLOTS_PER_EPOCH // TARGET_COMMITTEE_SIZE,
 //    ))
-func CommitteeCountAtSlot(state *pb.BeaconState, slot uint64) (uint64, error) {
-	epoch := SlotToEpoch(slot)
-	count, err := ActiveValidatorCount(state, epoch)
-	if err != nil {
-		return 0, errors.Wrap(err, "could not get active count")
-	}
-	var committeePerSlot = count / params.BeaconConfig().SlotsPerEpoch / params.BeaconConfig().TargetCommitteeSize
+func CommitteeCountAtSlot(validatorCount uint64) uint64 {
```
Comment (0xKiwi) [inline beacon-chain/core/helpers/committee.go]: Maybe change this to `activeValidatorCount`?

File: beacon-chain/blockchain/chain_info.go [hunk 2]
```diff
@@ -102,6 +105,16 @@ func (s *Service) HeadState(ctx context.Context) (*pb.BeaconState, error) {
 	return proto.Clone(s.headState).(*pb.BeaconState), nil
 }
 
+// HeadValidators returns a list of active validator indices from the head view of a given epoch.
+func (s *Service) HeadValidators(epoch uint64) ([]uint64, error) {
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/blockchain/chain_info.go]: This method should also describe that it only returns the active indices, not all of the validator indices.

File: beacon-chain/core/helpers/attestation.go [hunk 3]
```diff
@@ -121,21 +121,19 @@ func SlotSignature(state *pb.BeaconState, slot uint64, privKey *bls.SecretKey) (
...
 //   def is_aggregator(state: BeaconState, slot: Slot, index: CommitteeIndex, slot_signature: BLSSignature) -> bool:
 //    committee = get_beacon_committee(state, slot, index)
 //    modulo = max(1, len(committee) // TARGET_AGGREGATORS_PER_COMMITTEE)
 //    return bytes_to_int(hash(slot_signature)[0:8]) % modulo == 0
-func IsAggregator(state *pb.BeaconState, slot uint64, index uint64, slotSig []byte) (bool, error) {
-	committee, err := BeaconCommittee(state, slot, index)
-	if err != nil {
-		return false, err
-	}
+func IsAggregator(committeeCount uint64, slot uint64, index uint64, slotSig []byte) (bool, error) {
```
Comment (prestonvanloon) [inline beacon-chain/core/helpers/attestation.go]: Nice

File: beacon-chain/core/helpers/committee.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/core/helpers/committee.go]: Yeah.. what is this?

**Pull request merged**